File: sdks/python/opencomputer/mounts.py
"""FUSE-backed remote filesystem mounts inside a sandbox.

Mounts use ``rclone mount`` under the hood — one driver covering ~40 backends
(S3, GCS, Azure Blob, SFTP, WebDAV, Dropbox, etc.). Credentials are passed
inline, written to a tmpfs file inside the VM (mode 0600), and never persisted
on the worker. v1 does NOT auto-restore mounts on hibernate/wake — callers
re-issue ``add(...)`` after a wake if they need the mount back.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

import httpx

MountBackend = Literal["s3", "gcs", "azureblob", "sftp", "webdav", "dropbox"]


@dataclass
class MountInfo:
    """An active mount as tracked by the worker."""

    path: str
    remote: str
    read_only: bool
    backend: str = ""


@dataclass
class Mounts:
    """Mount remote filesystems via rclone+FUSE inside the sandbox."""

    _client: httpx.AsyncClient
    _sandbox_id: str

    async def add(
        self,
        path: str,
        remote: str,
        backend: MountBackend | None = None,
        creds: dict[str, str] | None = None,
        rclone_config: str | None = None,
        read_only: bool = True,
        mount_options: list[str] | None = None,
    ) -> MountInfo:
        """Mount a remote filesystem at ``path`` inside the sandbox.

        Args:
            path: Absolute path inside the VM where the remote will be mounted.
            remote: rclone remote spec — ``"<name>:<path>"`` (e.g. ``"s3:my-bucket/prefix"``).
            backend: One of ``s3``, ``gcs``, ``azureblob``, ``sftp``, ``webdav``,
                ``dropbox``. Determines how ``creds`` are templated into the
                rclone config. Omit when passing ``rclone_config`` directly.
            creds: Backend-specific config keys (rclone field names — e.g. for
                S3: ``access_key_id``, ``secret_access_key``, ``region``).
            rclone_config: Raw rclone config string. Overrides ``backend`` and
                ``creds`` — useful for backends not in the typed list or for
                advanced tuning.
            read_only: Default ``True``. Object-store FUSE mounts have
                well-known write footguns; opt in to RW explicitly.
            mount_options: Extra args appended to ``rclone mount`` (e.g.
                ``["--dir-cache-time", "1m"]``).
        """
        body: dict[str, object] = {
            "path": path,
            "remote": remote,
            "readOnly": read_only,
        }
        if backend is not None:
            body["backend"] = backend
        if creds is not None:
            body["creds"] = creds
        if rclone_config is not None:
            body["rcloneConfig"] = rclone_config
        if mount_options is not None:
            body["mountOptions"] = mount_options

        resp = await self._client.post(
            f"/sandboxes/{self._sandbox_id}/mounts", json=body
        )
        resp.raise_for_status()
        data = resp.json()
        return MountInfo(
            path=data["path"],
            remote=data["remote"],
            backend=data.get("backend", ""),
            read_only=data.get("readOnly", True),
        )

    async def list(self) -> list[MountInfo]:
        """List the mounts this worker is tracking for the sandbox.

        Returns empty after hibernate/wake — re-issue ``add()`` for any mounts
        you need back.
        """
        resp = await self._client.get(f"/sandboxes/{self._sandbox_id}/mounts")
        resp.raise_for_status()
        data = resp.json() or []
        return [
            MountInfo(
                path=entry["path"],
                remote=entry["remote"],
                backend=entry.get("backend", ""),
                read_only=entry.get("readOnly", True),
            )
            for entry in data
        ]

    async def remove(self, path: str) -> None:
        """Unmount a path previously passed to ``add()``. No-op if not mounted."""
        try:
            resp = await self._client.delete(
                f"/sandboxes/{self._sandbox_id}/mounts", params={"path": path}
            )
            if resp.status_code == 404:
                return
            resp.raise_for_status()
        except httpx.HTTPError:
            raise

File: sdks/typescript/src/mounts.ts
/**
 * Backends supported by the typed `creds` shape. For any backend not in this
 * list (or for advanced rclone tuning), pass `rcloneConfig` instead — the raw
 * string is dropped into the in-VM config file unchanged.
 */
export type MountBackend =
  | "s3"
  | "gcs"
  | "azureblob"
  | "sftp"
  | "webdav"
  | "dropbox";

export interface AddMountOpts {
  /** Absolute path inside the sandbox where the remote will be mounted. */
  path: string;
  /** rclone remote spec — `<name>:<path>` (e.g. `"s3:my-bucket/prefix"`). */
  remote: string;
  /**
   * Backend type. Determines how `creds` are templated into the rclone
   * config. Omit and pass `rcloneConfig` directly for backends not listed.
   */
  backend?: MountBackend;
  /**
   * Backend-specific credential / config keys (rclone config field names —
   * e.g. for S3: `access_key_id`, `secret_access_key`, `region`).
   *
   * Creds are written to a tmpfs file inside the VM (mode 0600) and never
   * persisted on the worker. v1 does not auto-restore mounts on hibernate —
   * re-call `mounts.add` after wake if you need the mount back.
   */
  creds?: Record<string, string>;
  /**
   * Raw rclone config to use verbatim. Overrides `backend`+`creds`. Useful
   * for backends not in the typed list, or for advanced tuning.
   */
  rcloneConfig?: string;
  /** Default `true`. Object-store FUSE mounts have well-known write footguns. */
  readOnly?: boolean;
  /** Extra args appended to `rclone mount` (e.g. `["--dir-cache-time", "1m"]`). */
  mountOptions?: string[];
}

export interface MountInfo {
  path: string;
  remote: string;
  backend?: string;
  readOnly: boolean;
}

export class Mounts {
  constructor(
    private apiUrl: string,
    private apiKey: string,
    private sandboxId: string,
    private token: string = "",
  ) {}

  private get headers(): Record<string, string> {
    if (this.token) return { "Authorization": `Bearer ${this.token}` };
    return this.apiKey ? { "X-API-Key": this.apiKey } : {};
  }

  /**
   * Mount a remote filesystem via rclone+FUSE inside the sandbox.
   *
   * @example
   * ```ts
   * await sandbox.mounts.add({
   *   path: "/mnt/data",
   *   remote: "s3:my-bucket",
   *   backend: "s3",
   *   creds: { access_key_id: "...", secret_access_key: "...", region: "us-east-1" },
   * });
   * ```
   */
  async add(opts: AddMountOpts): Promise<MountInfo> {
    const resp = await fetch(
      `${this.apiUrl}/sandboxes/${this.sandboxId}/mounts`,
      {
        method: "POST",
        headers: { "Content-Type": "application/json", ...this.headers },
        body: JSON.stringify(opts),
      },
    );
    if (!resp.ok) {
      const text = await resp.text();
      throw new Error(`Failed to add mount: ${resp.status} ${text}`);
    }
    return resp.json();
  }

  /**
   * List the mounts this worker is tracking for the sandbox. Returns empty
   * after hibernate/wake — re-issue `add()` for any mounts you need back.
   */
  async list(): Promise<MountInfo[]> {
    const resp = await fetch(
      `${this.apiUrl}/sandboxes/${this.sandboxId}/mounts`,
      { headers: this.headers },
    );
    if (!resp.ok) {
      const text = await resp.text();
      throw new Error(`Failed to list mounts: ${resp.status} ${text}`);
    }
    return resp.json();
  }

  /** Unmount a path previously passed to `add()`. No-op if not mounted. */
  async remove(path: string): Promise<void> {
    const resp = await fetch(
      `${this.apiUrl}/sandboxes/${this.sandboxId}/mounts?path=${encodeURIComponent(path)}`,
      { method: "DELETE", headers: this.headers },
    );
    if (!resp.ok && resp.status !== 404) {
      const text = await resp.text();
      throw new Error(`Failed to remove mount: ${resp.status} ${text}`);
    }
  }
}
